src/job-cache/job-cache.h

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright 2022 SiFive, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You should have received a copy of LICENSE.Apache2 along with
+ * this software. If not, you may obtain a copy at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// the `Cache` class provides the full interface
+// the the underlying complete cache directory.
+// This requires interplay between the file system and
+// the database and must be carefully orchestrated. This
+// class handles all those details and provides a simple
+// interface.
+
+#pragma once
+
+#include <errno.h>
+#include <fcntl.h>
+#include <json/json5.h>
+#include <sqlite3.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/file.h>
+#include <sys/ioctl.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <util/execpath.h>
+#include <util/mkdir_parents.h>
+#include <wcl/filepath.h>
+#include <wcl/trie.h>
+#include <wcl/xoshiro_256.h>
+
+#include <algorithm>
+#include <future>
+#include <iostream>
+#include <map>
+#include <random>
+#include <string>
+#include <thread>
+#include <unordered_map>
+#include <vector>
+
+#include "bloom.h"
+#include "logging.h"
+
+namespace job_cache {
+
+struct CachedOutputFile {
+  std::string path;
+  Hash256 hash;
+  mode_t mode;
+};
+
+struct JobOutputInfo {
+  std::string stdout_str;
+  std::string stderr_str;
+  int ret_code;
+  double runtime, cputime;
+  uint64_t mem, ibytes, obytes;
+};
+
+struct MatchingJob {
+  int64_t job_id;
+  std::vector<CachedOutputFile> output_files;
+  std::vector<std::string> input_files;
+  std::vector<std::string> input_dirs;
+  JobOutputInfo output_info;
+
+  JAST to_json() const {
+    JAST out(JSON_OBJECT);
+    JAST output_files_json(JSON_ARRAY);
+    JAST input_files_json(JSON_ARRAY);
+    JAST input_dirs_json(JSON_ARRAY);
+
+    // List the output files
+    for (const auto &output_file : output_files) {
+      // TODO: It's a shame we're throwing the hash out here :(
+      output_files_json.add(output_file.path);
+    }
+
+    // List the input files
+    for (const auto &input_file : input_files) {
+      input_files_json.add(input_file);
+    }
+
+    // List the input dirs
+    for (const auto &input_dir : input_dirs) {
+      input_dirs_json.add(input_dir);
+    }
+
+    // Add all the fields
+    out.add("output_files", std::move(output_files_json));
```

Comment:
I mean I can pull them out into constexpr variables if you'd like, that's not too big of a deal.

As for the "all json objects are represented by some class that does nothing but serialize/de-serialie" that is in fact already the reality in this change.

Reply:
constexpr would be nice since I think you reference these keys in other parts of the code. 

> that is in fact already the reality in this change

Ahh I didn't fully write out what I meant, that's by bad. Things like `output_files` should only ever be read via a compiler checked function like `output_files()` instead of a stringly typed process like `json["output_files"]` or `json[kOutputFiles]`

specifically to say, whenever the data shape of a map-like data object is known, I always prefer a class that only lets you access the fields via compile-time checked functions

Reply:
There are 3 json schemas

FindJobRequest (what the wake code sends the C++ code to find a job in the cache)
AddJobRequest (what the wake code sends the C++ code to add a job to the cache)
MatchingJob (what the C++ responds to a FindJobRequest with)

FindJobRequest only has a method for `json -> FindJobRequest`
AddJobRequest only has a method for `json -> AddJobRequest`
MatchingJob only has a method for `MatchingJob -> json`

I'm really not sure how much more strict I could get about it than that.